private/astarte_linked_list.h

```diff
@@ -0,0 +1,134 @@
+/*
+ * (C) Copyright 2018-2023, SECO Mind Srl
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR Apache-2.0
+ */
+
+/**
+ * @file astarte_linked_list.h
+ * @brief Utility module containing a doubly linked list implementation.
+ *
+ * @details This library does not perform deep copies when storing values.
+ * The user should ensure that any value is freed correctly when appropriate.
+ */
+
+#ifndef _ASTARTE_LINKED_LIST_H_
+#define _ASTARTE_LINKED_LIST_H_
+
+#include <stdbool.h>
+#include <stdlib.h>
+
+#include "astarte.h"
+
+typedef struct
+{
+    struct astarte_linked_list_node *head;
+    struct astarte_linked_list_node *tail;
+} astarte_linked_list_handle_t;
+
+typedef struct
+{
+    astarte_linked_list_handle_t *handle;
+    struct astarte_linked_list_node *node;
+} astarte_linked_list_iterator_t;
+
+/**
+ * @brief Initializes a new empty linked list
+ *
+ * @return A handle to the newly initialized list
+ */
+astarte_linked_list_handle_t astarte_linked_list_init(void);
+
+/**
+ * @brief Checks if the linked list is empty
+ *
+ * @param[in] handle Linked list handle
+ * @return true if the list is empty, false otherwise
+ */
+bool astarte_linked_list_is_empty(astarte_linked_list_handle_t *handle);
+
+/**
+ * @brief Append an item to the end of a linked list
+ *
+ * @param[inout] handle Linked list handle
+ * @param[in] value Item to append to the linked list
+ * @return One of the follwing error codes:
+ * - ASTARTE_ERR_OUT_OF_MEMORY if memory allocation failed,
+ * - ASTARTE_OK if operation has been successful
+ */
+astarte_err_t astarte_linked_list_append(astarte_linked_list_handle_t *handle, void *value);
+
+/**
+ * @brief Remove and return the last item from a linked list
+ *
+ * @param[inout] handle Linked list handle
+ * @return The removed item, will return NULL when no item is present in the list
+ */
+void *astarte_linked_list_remove_tail(astarte_linked_list_handle_t *handle);
```

Comment:
@harlem88 I couldn't find a better solution than returning a pointer that might be NULL here. However, I think it kind of conflicts with the overall philosophy of this driver where errors are returned and evaluated.
The only other solution I could think of is:
```
astarte_err_t astarte_linked_list_remove_tail(astarte_linked_list_handle_t *handle, void **value);
```
However, I am not very happy with the ergonomics of double-pointers.
What do you think? Do you have any suggestions?

Reply:
I would like to keep the last version committed with the `astarte_err_t`.